### ISSUE TITLE
Information on corners and their connection with faces and cells

### ISSUE THREAD

**Issue by gregtucker** [landlab/landlab#458]
@mcflugen : Couple of questions on your Graph package:

1) Here, or elsewhere, is there a mechanism to compute the number of corners and their (x,y) locations? Or does that still need to be created?

2) Am I correct in understanding that if you had corner (x,y) locations, you could create a Graph with these as inputs, and it would find the corresponding "links" (really FACES) and "patches" (really CELLS)?

3) Or is there already somewhere a "complete" code to find the full set of geometry and topology for nodes-links-patches AND corners-faces-cells?

**Comment by mcflugen** [comment]
1. `graph.number_of_corners` gives the total number of corners. `x_of_corner` gives the x coordinate of each corner.

2. If I'm understanding correctly, this is not something you can do now. I need to think about this a bit more but, in general, I'm not sure it's possible. To do this, we may have to have special cases for each of our grid types. As an illustration of the problem, imagine you wanted to created a dual raster graph given a graph of `2 x 3` corners. I think you would expect the new graph to have `3 x 4` nodes, correct? However, to add the perimeter nodes (the ones outside of the corners), you have to know that they should be aligned with the same row and column spacing of the corners. I think a typical dual graph would actually place those nodes at infinity. I think adding these perimeter nodes will be a problem with any kind of graph.

3. Not yet. I created several Jupyter notebooks explaining how to use graphs but removed them from the tutorials repository as the branch hasn't been merged yet. Folks were trying to run them but were unable to because they were either using a release version of landlab or were on the `master` branch.

**Comment by gregtucker** [comment]
I tried corners but got the following:

`from landlab.graph import Graph

node_x, node_y = [0, 0, 0, 1, 1, 1, 2, 2, 2], [0, 1, 2, 0, 1, 2, 0, 1, 2]

graph = Graph((node_y, node_x))

graph.x_of_node
Out[4]: array([ 0.,  1.,  2.,  0.,  1.,  2.,  0.,  1.,  2.])

graph.x_of_corner
Traceback (most recent call last):

  File "<ipython-input-5-cc0c778e632a>", line 1, in <module>
    graph.x_of_corner

AttributeError: 'Graph' object has no attribute 'x_of_corner'`

**Comment by mcflugen** [comment]
@gregtucker: Ah, I didn't realize you were actually trying to use it! In that case, it might be easier to work with a rectilinear grid. The `Graph` and `DualGraph` classes are really just base classes to create other graphs from. Try the following,
```python
>>> from landlab.graph import DualUniformRectilinearGraph
>>> graph = DualUniformRectilinearGraph((3, 4), spacing=(1., 2.))
>>> graph.x_of_corner
array([ 1.,  3.,  5.,  1.,  3.,  5.])
>>> graph.y_of_corner
array([ 0.5,  0.5,  0.5,  1.5,  1.5,  1.5])
```
The classes that start with "Dual" are *node-link-patch* graphs along with their dual (*corner-face-cell*) graph. In my example, if you used a `UniformRectilinearGraph` you would get the same `AttributeError` you saw before. Since the dual graphs are what we will usually be using, perhaps a name change is in order.

I'll clean this up for our meeting next week and can show the group some examples if that would be of interest.

**Comment by gregtucker** [comment]
Thanks @mcflugen. This works very well, and I was able to write a JSON file-writer using a combination of GRID and GRAPH. However, one issue crops up, having to do with hex grids. It looks as some of the patches are in the wrong order. In the example below, patch 11 has the nodes (and links, it turns out) that actually belong to patch 13. Do you agree? I suppose this could happen if there is a slight round-off error in y coordinates, such that patch 13 has an ever-so-slightly smaller y coordinate than 11 or 12, though I don't know what that would occur.

>>> from landlab.graph import DualHexGraph
>>> hg = DualHexGraph((3, 4), spacing=10.0, node_layout='hex')
>>> for p in range(hg.number_of_patches):
...    print p, hg.nodes_at_patch[p]
0 [5 0 1]
1 [6 1 2]
2 [7 2 3]
3 [5 4 0]
4 [6 5 1]
5 [7 6 2]
6 [8 7 3]
7 [9 4 5]
8 [10  5  6]
9 [11  6  7]
10 [12  7  8]
11 [11 10  6] <== SHOULD BE 10 5 9 I THINK
12 [12 11  7] <== SHOULD BE 11 10 6
13 [10  9  5] <== SHOULD BE 12 11 7

**Comment by mcflugen** [comment]
@gregtucker, you are correct that there is an ordering problem. You are also correct about the cause of the problem. This is fixed in the version that's on my computer but not in the version that's on GitHub. I'll have to fix that today.

**Comment by gregtucker** [comment]
Aha, good to hear, thanks @mcflugen
